Fix warnings within useSelect of the bulk editing header

louwie17 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

- Creates a new hook in core data called `getEditedEntityRecords` which, retrieves multiple edited records.
     - This does trigger a `getEntityRecord` request for every record in the list.
- Updates the `PostActions` component to use `getEditedEntityRecords` and `getEntityRecordsPermissions` to avoid map usage within useSelect.
- Updates the `TemplateEdit` function to not use `filter` within `useSelect` causing warnings ( breaking memoization ).

## Why?

Avoids this `useSelect` warning as pointed out in this comment: https://github.com/WordPress/gutenberg/pull/67743#discussion_r1878398085 ( same goes for the template field ):

![394355180-4ea73568-1a97-4285-88b4-c5a07873ce71](https://github.com/user-attachments/assets/07a82e8e-c291-4fa7-bf51-de08f3191924)

## How?

Removing usage of `map`, `filter`, and empty array within a `useSelect`. 

## Testing Instructions
1. Enable the "Quick Edit in DataViews" experiment
2. Enable the 2024 or 2025 theme and go to **Appearance > Editor > Pages**
3. Change the view in the top right to **Table** and open the right side bar
4. Select multiple pages, it should show "2 pages" in the header and a 3 dot menu.
5. Open the console and make sure the `The 'useSelect; hook returns different values when called with the same state and parameters. This can lead to unnecessary re-renders and performance issues if not. ` is not being triggered for either `items` and `permissions` or `availableTemplates` and `templates`.
6. Select either the `homepage` page or the `posts page` and check if the templates field is either disabled, or doesn't show the `Swap template` option. ( You can set the `homepage` and `posts page` pages within **Settings > Reading > 
Your homepage displays**
7. Now edit a single page and select the `Page` tab in the editing settings menu ( on the far right )
8. And make sure the same error is not being triggered either.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

**Before**

https://github.com/user-attachments/assets/b4c0619c-d080-4743-90cf-43b312ba8e08

**After**

https://github.com/user-attachments/assets/35708ee3-5207-47b6-9d9f-5816d6295b1f

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

louwie17 committed:
Add getEditedEntityRecords selector

louwie17 committed:
Remove map usage from useSelect

louwie17 committed:
Remove filter usage from template useSelect

louwie17 commented on packages/core-data/src/selectors.ts:
```diff
@@ -891,6 +891,57 @@ export const getEditedEntityRecord = createSelector(
+		recordIds: EntityRecordKey[]
+	): Array< ET.Updatable< EntityRecord > | false > => {
+		return recordIds.map( ( recordId ) =>
+			getEditedEntityRecord( state, kind, name, recordId )
```
Given I re-use `getEditedEntityRecord`, it would trigger the `getEntityRecord` resolver for every record id in the list.
@youknowriad or @oandregal do you know why this happens when we already technically got all the records in the able using `useEntityRecords`. Is it because we are not triggering a `finishResolution` for each record as part of that call?

youknowriad replied:
> Is it because we are not triggering a finishResolution for each record as part of that call?

That's probably true.

What are our options? Should we use `getEntityRecords` instead of `getEntityRecord` in `getEditedEntityRecords`

louwie17 replied:
Actually I stand corrected, the `finishResolution` does work correctly. The requests I am seeing are `OPTIONS` calls triggered by the `canUser` selector for getting the permissions: https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/resolvers.js#L524-L533

Mamaduka replied:
We can run permissions selector after items are resolved. This way we can avoid dispatching a permission request for each item.

louwie17 replied:
I can try, but I am skeptical that would work. Given the individual record calls are not triggered because they are resolved by the `getEntityRecords` call. 
Cause would it be enough to check permissions just for the `postType`? or stick with checking for every record.

louwie17 replied:
Ok, so the problem was that I was passing the `postIds` in as strings into the permission call which re-triggered the requests. Although it had already been resolved prior by the data for the DataViews list.
I updated the above in this commit: https://github.com/WordPress/gutenberg/pull/67872/commits/6972c7969594019ce7ac61ba0a598fb4f0572156 to follow the same format as this hook: https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/hooks/use-entity-records.ts#L158-L204
Where we get the records first, and then get the list of ids from the actual records ( instead of relying on the parsed ids ). This seems to fix the trigger for additional `OPTIONS` request.

Mamaduka replied:
It should. The `getEntityRecords` selector will try to resolve permissions for each entity when possible (ref #64504). However, we had a race condition because it called both selectors simultaneously.

> Ok, so the problem was that I was passing the postIds in as strings into the permission call which re-triggered the requests. 

Maybe we should normalize IDs before passing them to selectors.

louwie17 replied:
> Maybe we should normalize IDs before passing them to selectors.

Yeah not a bad plan, with the current changes I suppose we are doing a form of normalizing by relying on the ID type that the entity record has.

louwie17 commented on packages/fields/src/fields/template/template-edit.tsx:
```diff
@@ -39,52 +39,49 @@ export const TemplateEdit = ( {
+			).getEntityRecords< WpTemplate >( 'postType', 'wp_template', {
+				per_page: -1,
+				post_type: postType,
+			} );
```
Remove the `?? []` and handle this outside of the `useSelect`.

louwie17 commented on packages/fields/src/fields/template/template-edit.tsx:
```diff
@@ -39,52 +39,49 @@ export const TemplateEdit = ( {
-				postType === 'page' && getHomePage()?.postId === +postId;
-
-			const allowSwitchingTemplate = ! isPostsPage && ! isFrontPage;
+				postType === 'page' && +getHomePage()?.postId === postId;
```
Small fix, the `getPostsPageId()` and `getHomePage()?.postId` actually return strings, so `allowSwitchingTemplate` always returned true.

youknowriad replied:
Should we make `getHomePage` return the right type?

louwie17 replied:
Meaning an integer? It seems like in some of the other logic you were already depending on it being a string: https://github.com/WordPress/gutenberg/blob/4867dfc6bb9286865ae19e332b06a29b35ce7ab1/packages/core-data/src/private-selectors.ts#L188
Similar scenario with the `getPostsPageId` that explicitly converts it to a string: https://github.com/WordPress/gutenberg/blob/4867dfc6bb9286865ae19e332b06a29b35ce7ab1/packages/core-data/src/private-selectors.ts#L115

Instead of converting it to an integer above, maybe I should just keep the `postId` as a string?

youknowriad replied:
If I'm not wrong, the "posts" REST API returns post Ids as integers, except for templates and template Parts. So IMO, we should stay consistent with that. When the post type is "template" or "template part" use string ids, otherwise integer ids.

There might be some type casting we need to do at the route levels, because urls always give strings. 

WDYT?

Mamaduka replied:
> If I'm not wrong, the "posts" REST API returns post Ids as string

@youknowriad, do you mean integers? REST API should return DB ID values as ints.

youknowriad replied:
yes sorry :)

louwie17 replied:
Yeah I agree we should stay consistent, and happy to stick with `When the post type is "template" or "template part" use string ids, otherwise integer ids.`.
Would it help to have a `sanitize` function for this that includes the `postType` and uses the above logic?

Things may get a bit confusing in functions like  `getHomePage` ( https://github.com/WordPress/gutenberg/blob/4867dfc6bb9286865ae19e332b06a29b35ce7ab1/packages/core-data/src/private-selectors.ts#L144-L156 ) where it returns either a `page` or `wp_template` so the `postId` will likewise be either a string or integer. Which in a way I see why you piped it through `normalizePageId` to convert it to a string.

louwie17 replied:
I updated the above function to use integers when dealing with pages or posts in this commit: https://github.com/WordPress/gutenberg/pull/67872/commits/ba6868a3a10c615c20c1bc53f9c53f74db49f024
I also converted the list of ids coming from the router to integers, as part of this commit: https://github.com/WordPress/gutenberg/pull/67872/commits/9c55bef36e3190e74e5864f0b3a56989b12f7509

louwie17 committed:
Update hook to use correct ids for permissions

louwie17 committed:
Compare using strings for template id

Mamaduka commented on packages/core-data/src/selectors.ts:
```diff
@@ -891,6 +891,57 @@ export const getEditedEntityRecord = createSelector(
+ *
+ * @return The list of entity records, merged with their edits.
+ */
+export const getEditedEntityRecords = createSelector(
```
Let's make this new selector private for now.

louwie17 replied:
Done: https://github.com/WordPress/gutenberg/pull/67872/commits/e4cd460ee05bb514677af29f12a1492721960843

louwie17 committed:
Move new selector to private

louwie17 commented:
@Mamaduka & @youknowriad this is ready for a re-review. I moved the selector to be private and updated the `useEditedEntityRecordsWithPermissions` hook to avoid the permission `OPTION` call triggers.
I stuck with using `getEditedEntityRecord` within the `getEditedEntityRecords`, as this did avoid triggering requests, given each record was already resolved by the data from `Dataviews`. Of-course this would not be the case for every use case.

Mamaduka commented:
Thanks, @louwie17!

Can you give me the steps to reproduce the issue fixed by 6972c7969594019ce7ac61ba0a598fb4f0572156? I think splitting data retrieval into multiple `useSelect` calls shouldn't be necessary.

louwie17 commented:
> Can you give me the steps to reproduce the issue fixed by [6972c79](https://github.com/WordPress/gutenberg/commit/6972c7969594019ce7ac61ba0a598fb4f0572156)? I think splitting data retrieval into multiple `useSelect` calls shouldn't be necessary.

It fixed the multiple `OPTIONS` requests for each selected page, that was triggered by the permissions call.
Splitting the `useSelect` up does feel a bit tedious, I followed the exact same form as here: https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/hooks/use-entity-records.ts#L158-L204

I suppose I can move pretty well everything into a single `useSelect`. I didn't want to just normalize the IDs to integers right away, in the case where we re-use this for the templates, posts, and pattern screens ( given the id's are not always integers ).
But let me see if I can clean it up a bit into a single `useSelect`.

louwie17 committed:
Move getEntityConfig inside of other useSelect

louwie17 commented:
> Can you give me the steps to reproduce the issue fixed by [6972c79](https://github.com/WordPress/gutenberg/commit/6972c7969594019ce7ac61ba0a598fb4f0572156)? I think splitting data retrieval into multiple `useSelect` calls shouldn't be necessary.

I was not able to move everything into a single `useSelect` as this means that we generate the `ids` list with a `map` inside of the `useSelect` causing the `getEntityRecordsPermissions` selector to be re-triggered. 
I did move the `entityConfig` call into one of the other `useSelects` as it wasn't necessary for it to be in a different one.

louwie17 committed:
Update post/page related functions to return ID as number

louwie17 committed:
Handle post ids as numbers

louwie17 commented:
@Mamaduka for when you have some time, this PR is ready for a re-review :)

talldan commented on packages/editor/src/components/post-actions/index.js:
```diff
@@ -21,27 +21,48 @@ import { usePostActions } from './actions';
 		[ postIds, postType ]
 	);
 
+	const ids = useMemo(
+		() =>
+			items?.map( ( record ) => record[ entityConfig?.key ?? 'id' ] ) ??
+			[],
+		[ items, entityConfig?.key ]
+	);
+
+	const permissions = useSelect(
+		( select ) => {
+			const { getEntityRecordsPermissions } = unlock(
+				select( coreStore )
+			);
+			return getEntityRecordsPermissions( 'postType', postType, ids );
+		},
+		[ ids, postType ]
+	);
```
This is a bit of a drive-by comment, and I don't know much about the code, so I might be far off the mark.

Can `getEntityRecordsPermissions` do this whole dance with the keys internally?

It seems inconsistent that it's possible to call `getEditedEntityRecords` with the `postIds`,  but for `getEntityRecordsPermissions` the `postIds` have to be mapped to a particular key. It might also be a gotcha for any users of the API.

louwie17 replied:
> Can getEntityRecordsPermissions do this whole dance with the keys internally?

It certainly could, I think we did have to create a separate selector to handle that.
I currently followed the same format as an already existing hook: https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/hooks/use-entity-records.ts#L158-L204

> It seems inconsistent that it's possible to call getEditedEntityRecords with the postIds, but for getEntityRecordsPermissions the postIds have to be mapped to a particular key. It might also be a gotcha for any users of the API.

Yeah, I can probably remove it now given I explicitly converted the `postIds` to integers upstream: https://github.com/WordPress/gutenberg/pull/67872/files/9c55bef36e3190e74e5864f0b3a56989b12f7509#diff-4ed486e973856f952d2b80aea381fc1f5a1d0004995a0b68554147b02cee0b55R36-R43

talldan commented on packages/editor/src/components/post-actions/index.js:
```diff
@@ -21,27 +21,48 @@ import { usePostActions } from './actions';
-			permissions: permissions[ index ],
-		} ) );
+		return (
+			items?.map( ( item, index ) => ( {
+				...item,
+				permissions: permissions ? permissions[ index ] : undefined,
+			} ) ) ?? []
```
Again I'll leave a caveat that I don't know the code very well, but I feel a bit nervous about the way the permissions are merged into the entity records here.

It might be better if this was TypeScript, but because it's JS, other contributors will have to trace back through the code to find out whether the items contain the permissions key or not. It's the kind of thing that could lead to bugs if another contributor makes a bad assumption.

I think it's clearer and more intentional if they're kept separate, and also the hook ends up returning a more pure form of `getEntityRecordPermissions`, which feels more elegant to me. I think it could simplify the code quite a bit.

louwie17 replied:
> It might be better if this was TypeScript, but because it's JS, other contributors will have to trace back through the code to find out whether the items contain the permissions key or not. It's the kind of thing that could lead to bugs if another contributor makes a bad assumption.

I don't know if I totally agree with this, given we explicitly set this list to the `itemsWithPermissions` constant, which should give enough context.

Having said that, I do agree that keeping them separate would be ideal. This would mean either updating the `isEligible` ( [link](https://github.com/WordPress/gutenberg/blob/2e3e6e40fc8dac90a34bf880fe5749c638e4861d/packages/dataviews/src/types.ts#L427) ) function to take permissions, or push consumers to handle getting permissions themselves within the `isEligible` function. This may mean we should support it being `async` as well.
@youknowriad or @oandregal curious if you have an opinion here?
I would lean towards the former, where we may add an additional `context` param to the `isEligible` function that may contain permissions.
Advocating for a `context` param, as permissions is post specific and would not make sense for some of the other use cases.

talldan commented on packages/editor/src/components/post-actions/index.js:
```diff
@@ -21,27 +21,48 @@ import { usePostActions } from './actions';
+			const entityRecords = getEditedEntityRecords(
+				'postType',
+				postType,
+				postIds
```
At the start of `PostActions`, there's a `useMemo` call (I can't leave a code comment there as it's not part of the changeset):
```js
	const _postIds = useMemo( () => {
		if ( Array.isArray( postId ) ) {
			return postId;
		}
		return postId ? [ postId ] : [];
	}, [ postId ] );
```

I think that's unnecessary overhead, the coercion to an array can be performed in this `useSelect` within `useEditedEntityRecordsWithPermissions`. From what I could tell `useEditedEntityRecordsWithPermissions` is the only place `_postIds` is used, so I think it's a good idea to move that code inside this hook.

louwie17 replied:
Yeah that is also an option, although if we move the `useMemo` code into the `useSelect` it would mean we pass a new array of ids into `getEditedEntityRecords` which in turn will trigger the memoization of the `getEditedEntityRecords` selector and cause the same warning to appear that this PR attempts to prevent ( when handling single ids ).

Mamaduka commented:
I'm considering reverting the bulk actions from the Quick Edit panel to avoid accidentally leaking this issue in the next major WP release, especially since Quick Edit for DataViews is still experimental.

I see this as an opportunity to re-evaluate bulk edits at a lower level. Our current hacks in the codebase don't scale well and add developer overhead.

* Can we simplify bulk actions capability checks and make them similar to the [WP List Table](https://github.com/WordPress/wordpress-develop/blob/88796678fd838ab0e790d22558f258c906262400/src/wp-admin/includes/class-wp-posts-list-table.php#L432-L453), where only general permissions are checked?
* Could we use the [REST API Batch framework](https://make.wordpress.org/core/2020/11/20/rest-api-batch-framework-in-wordpress-5-6/)?

@louwie17, thanks again for contributing, and hopefully, we can resume work on this enhancement shortly.

youknowriad commented:
> I'm considering reverting the bulk actions from the Quick Edit panel 

What exactly are you considering reverting since the whole panel is experimental so unless I'm missing something is not going to leak?

Mamaduka commented:
Probably #67743 and #67390. I haven't started working on them yet, but I will try to keep the changes to a minimum.

> whole panel is experimental so unless I'm missing something is not going to leak?

The `PostActions` component isn't experimental and is used by editors. Bulk editing changes for it aren't "stable" or used/useful outside experimental Quick Edit.

youknowriad commented:
> The PostActions component isn't experimental and is used by editors.

I think making PostActions component support bulk actions is a good thing, I don't really see why we'd need to revert that personally. Your call though.

louwie17 commented:
> * Can we simplify bulk actions capability checks and make them similar to the [WP List Table](https://github.com/WordPress/wordpress-develop/blob/88796678fd838ab0e790d22558f258c906262400/src/wp-admin/includes/class-wp-posts-list-table.php#L432-L453), where only general permissions are checked?

Doing something similar as this makes a lot more sense IMO. Cause is it actually possible to have different permissions for specific posts or pages? For example could a user have edit/delete permissions on one post and not the other?
Or would this only apply to templates?

Is this something we already have a data store selector for, or more generally an API that provides this on a post type basis?

> * Could we use the [REST API Batch framework](https://make.wordpress.org/core/2020/11/20/rest-api-batch-framework-in-wordpress-5-6/)?

Are we not already using this? Currently the bulk editing is handled locally and the user has to trigger the save afterwards, that is all handled by the entity data store ( as far as I am aware ). This could definitely use the batch approach if there are multiple records that need saving.
This can be handled outside of the DataViews scope though, I did have to dig in a bit deeper to see how this is currently handled.   

> @louwie17, thanks again for contributing, and hopefully, we can resume work on this enhancement shortly.

👍 

> Probably https://github.com/WordPress/gutenberg/pull/67743 and https://github.com/WordPress/gutenberg/pull/67390. I haven't started working on them yet, but I will try to keep the changes to a minimum.
> The PostActions component isn't experimental and is used by editors. Bulk editing changes for it aren't "stable" or used/useful outside experimental Quick Edit.

Let me know if you need a review on this. I believe the `PostActions` in the editor is always a single post, so hopefully it won't require a full revert. Unless you think getting this PR over the line would suffice, in that case feel free to take it over if need to be.

t-hamano commented:
This PR still looks like it's under discussion, but I think at least the browser warning should be fixed in the WP 6.8 release.

I'll add this PR to the project board so we don't forget to take action.

Mamaduka commented:
Sorry, @louwie17! I didn't have the resources to help move this forward. Do you have availability to revert the required changes from https://github.com/WordPress/gutenberg/pull/67743 and https://github.com/WordPress/gutenberg/pull/67390?

Mamaduka commented:
Thanks for the reply, @louwie17! Yeah, right now, it's impossible to clean revert using Git, so it should be a manual effort.

I think this PR still needs polishing. Here's a small example: The `getEditedEntityRecords` implementation should be similar to `getEntityRecords` and use a forwarded resolver, just like `getEditedEntityRecord`.

Also, there's always a chance we discover minor issues with this feature, and I'm not confident in our current resources to follow up properly in time.
